### ISSUE TITLE
NoSuchMethodError in RedisStateMachine.readLong

### ISSUE THREAD

**Issue by chscorpio** [mp911de/lettuce#322]
While using lettuce version 4.2.1.Final and netty version 4.1.0.Final, I get a NoSuchMethodError in RedisStateMachine.readLong.

See the stacktrace in the gist:
https://gist.github.com/chscorpio/ce3af686589eb5f49e3dc30102e9e637

It seems this issue was already identified in https://github.com/mp911de/lettuce/issues/274 and fixed on master in https://github.com/mp911de/lettuce/commit/824618dc442c1eb7e0e3bb8078a85cdff750a126 but not included in 4.2.1.Final.

Can this fix or a variation of it be fixed for a 4.2.2.Final or subsequent release? Are there any workarounds that will allow me to continue using netty 4.1.0?

The reason this is an issue for us now is that we decided to upgrade to 4.2.1.Final from 4.0.0.Final after we encountered reconnection issues which we see resolved for 4.2.1.Final. The good news is that our QA team has verified that 4.2.1.Final has fixed the reconnection issues and now we want to update all our services to use Lettuce 4.2.1.Final. Unfortunately, some of our services also depend on grpc 0.13.2 which depends on netty 4.1.0 and hence this NoSuchMethodError is blocking us from upgrading those services.

**Comment by chscorpio** [comment]
I just realized that Lettuce 4.2.1.Final tries to deal with the differences in the RedisStateMachine constructor with the following code...

```
        Version nettyBufferVersion = Version.identify().get("netty-buffer");
        if(nettyBufferVersion != null) {
            useNetty41ByteBufCompatibility = nettyBufferVersion.artifactVersion().startsWith("4.1");
        } else {
            useNetty41ByteBufCompatibility = false;
        }
```

I'll be debugging now why useNetty41ByteBufCompatibility ends up false for our service.

**Comment by chscorpio** [comment]
Found it! The netty-buffer version isn't in our shaded jar because META-INF/io.netty.versions.properties is overwritten by netty-codec.

I need to find a way to merge all the META-INF/io.netty.versions.properties into our shaded jar so that the netty-codec and netty-buffer versions are found.

**Comment by chscorpio** [comment]
Adding the following lines to the maven-shade-plugin definition in our pom.xml worked for me...

```
<transformer implementation="org.apache.maven.plugins.shade.resource.AppendingTransformer">
    <resource>META-INF/io.netty.versions.properties</resource>
</transformer>
```

See details at https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#AppendingTransformer

**Comment by gjesse** [comment]
I'm not seeing this as resolved - in fact as far as i can see the fix #274 never made it into 4.2.1.Final or 4.2.2.Final.

Note screenshot from 4.2.2.Final-sources.jar: https://www.dropbox.com/s/62euk5a1oyk1ydm/Screenshot%202016-08-29%2012.30.20.png?dl=0

here's the file in the 4.2.x branch (note use of ToLongProcessor)

https://github.com/mp911de/lettuce/blob/4.2.x/src/main/java/com/lambdaworks/redis/protocol/RedisStateMachine.java#L50

and the fix here - in master:

https://github.com/mp911de/lettuce/blob/master/src/main/java/com/lambdaworks/redis/protocol/RedisStateMachine.java#L50

Also, here's the io.netty.versions.properties for my final application jar:

```
netty-codec-http.version=4.1.4.Final
netty-codec-http.buildDate=2016-07-27 13\:37\:22 +0200
netty-codec-http.commitDate=2016-07-27 13\:16\:59 +0200
netty-codec-http.shortCommitHash=9466b32
netty-codec-http.longCommitHash=9466b32d05ecf63b7e3a16e8e432ca5def21dc1e
netty-codec-http.repoStatus=clean
```

**Comment by gjesse** [comment]
my bad. i misunderstood the fix and this issue. plz ignore my comment.
